Multiversion support

calvellido opened the pull request:
* Set a new `makeMultiversionMicrosite` task.
* Add new `micrositeVersionList` setting.
* `sbt` tasks reorganization in order to make way for the new task, while keeping same behaviour instact.
* Add a bunch of tasks new auxiliary functions.
* Include logic necessary (layout/JS) for when a multiversion selector is included.
* Set proper styling for the version selector. 
* Update docs.

This closes #116

calvellido commented:
This is ready to rumble now :rocket:

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+  def pluginProjects(pluginName: String): Option[Array[String]] = {
+    val sbtPluginsOutput = "sbt --error plugins".lineStream
+    val pluginLine =
+      sbtPluginsOutput.find(line => line.trim.startsWith(s"$pluginName: enabled in "))
```
```suggestion
      sbtPluginsOutput.find(_.trim.startsWith(s"$pluginName: enabled in "))
```

calvellido replied:
Ha!  And I was asking myself how could I use that underscore syntax... 😅

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+
+  def generateVersionList(versionStringList: List[String], ownVersion: String) = {
+    versionStringList
+      .map(version => {
```
The braces wouldn't be needed

calvellido replied:
Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/22cdc14758ba83507df43b91a8b6189d20f155a1

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+    val jekyllDir  = "jekyll"
+    val targetPath = s"$targetDir$jekyllDir/_data/versions.json"
+    createFile(targetPath)
+    writeContentToFile(content.asJson.toString, targetPath)
```
You might want to write the json with a circe Printer:

```scala
content.asJson.pretty(Printer.noSpaces) // Or
content.asJson.pretty(Printer.spaces2)
```

calvellido replied:
Just curious, what is the benefit?  This is the format being printed now:

```json
[
  {
    "name" : "jv-116-multiversion",
    "own" : true
  },
  {
    "name" : "v0.9.7",
    "own" : false
  }
]
```

I guess `asJson` is already using `spaces2` internally?

https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Json.scala#L188

juanpedromoreno replied:
Yup, it seems like :) . Thanks for checking.

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+    pluginLine match {
+      case Some(line) => Some(line.trim.stripPrefix(s"$pluginName: enabled in ").split(", "))
+      case None       => None
+    }
```
```scala
pluginLine.map(_.trim.stripPrefix(s"$pluginName: enabled in ").split(", "))
```

?

calvellido replied:
Would the use of `map` there return a `None` in the case that `pluginLine` was one?  I was under the assumption that `map` would only work with _right_ values, but I wasn't sure what would happen for `None` value.

calvellido replied:
Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/41e996fd7cf2934d9ab7d07d58d722506ab70b9b

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+      baseUrl: String,
+      version: String): Unit = {
+    val newBaseUrl =
+      if (version != "") s"${baseUrl}/$version" else s"${baseUrl}"
```
```suggestion
      if (version.nonEmpty) s"$baseUrl/$version" else baseUrl
```

juanpedromoreno commented on src/main/scala/microsites/MicrositeKeys.scala:
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
+          s"""clean; set ${projects(0)}/micrositeBaseUrl := "$newBaseUrl"; ${projects(0)}/makeMicrosite""").!
+        Files.move(
+          FileSystems.getDefault().getPath(sourceDir),
+          FileSystems.getDefault().getPath(s"$targetDir/$version"),
```
I believe you can use:

```scala
Paths.get(pathString)
```

in both cases above.

calvellido replied:
Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/41e996fd7cf2934d9ab7d07d58d722506ab70b9b

juanpedromoreno commented on docs/docs/docs/build-the-microsite.md:
```diff
@@ -12,7 +12,28 @@ Once you have written your documents, you can build the microsite running this s
+
+# Multiversion support
+
+Starting on version `1.0.0` you can build microsites including different versions of your library. This is available for the `light` theme, and you need to have [`git`](https://git-scm.com/) installed and available in your system. Most probably you already have it, but in other case please check your system docs, or go to the [git website](https://git-scm.com/download/) on how to proceed.
```
Do you think it could be possible to use this feature in other themes down the road?

calvellido replied:
Definitely!  It's only not being offered to the `pattern` theme at this time, along more features that are not ready yet, but I hope to bring them all over.

This PR was taking longer than I was expecting, so I reduce the scope a bit for this release.

juanpedromoreno replied:
Makes sense

calvellido commented:
@juanpedromoreno thanks for the review, I'll keep addressing the comments later.

Not specifically related with the review, but do you know if there is any aux function to delete a directory (empty or not) in `sbt-org-policies` (or any other dependency in the project for that matter)?  It could have been of help here (to delete the temporary directory where the microsites are managed until they are moved to their final destination), but I wasn't sure if I should be implementing it here.

juanpedromoreno commented:
@calvellido Maybe you could:

```scala
new scala.reflect.io.Directory(new java.io.File("/yourDirectory")).deleteRecursively()
```
